File: src/MailSearch.Core/Importer/PstImporter.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Threading.Tasks;
using MailSearch.Core.Database;
using MailSearch.Importer;


namespace MailSearch.Core.Importer;

/// <summary>
/// Imports PST/OST files into the MailSearch database.
///
/// PST/OST reading requires a third-party library. This implementation uses an
/// <see cref="IPstReader"/> abstraction so you can plug in any library.
///
/// Recommended free/open-source option:
///   XstReader.Api (https://github.com/iluvadev/XstReader) – add via NuGet and
///   implement <see cref="IPstReader"/> wrapping XstReader calls.
///
/// Commercial options: Aspose.Email, Independentsoft.Email.Mapi.
/// </summary>
public static class PstImporter
{
    /// <summary>
    /// Imports a PST/OST file using the supplied <paramref name="pstReader"/>.
    /// </summary>
    public static async Task<ImportResult> ImportPstAsync(
        string pstFilePath,
        MailSearchRepository repo,
        string attachmentDir,
        IPstReader pstReader,
        Action<int>? onProgress = null)
    {
        var result = new ImportResult();
        await pstReader.ReadMessagesAsync(pstFilePath, (msg) =>
        {
            try
            {
                var email = new Models.Email
                {
                    MessageId = msg.MessageId,
                    Subject = msg.Subject ?? "(no subject)",
                    From = new Models.EmailAddress
                    {
                        Name = msg.SenderName,
                        Email = msg.SenderEmail ?? string.Empty,
                    },
                    To = EmailNormalizer.ParseAddressList(msg.DisplayTo),
                    Cc = EmailNormalizer.ParseAddressList(msg.DisplayCc),
                    Bcc = EmailNormalizer.ParseAddressList(msg.DisplayBcc),
                    Date = msg.DeliveryTime,
                    BodyText = msg.BodyText,
                    BodyHtml = msg.BodyHtml,
                    HasAttachments = msg.HasAttachments,
                    AttachmentCount = msg.AttachmentCount,
                    SourceFile = Path.GetFullPath(pstFilePath),
                    ImportedAt = DateTime.UtcNow,
                    FolderPath = msg.FolderPath,
                };

                int emailId = repo.InsertEmail(email);
                if (emailId == 0)
                {
                    result.Skipped++;
                    return;
                }

                foreach (var att in msg.Attachments)
                {
                    if (string.IsNullOrEmpty(att.Filename) || att.Data == null) continue;
                    try
                    {
                        var attachmentRecord = EmailNormalizer.BuildAttachment(
                            emailId, att.Filename, att.MimeType, att.Data, attachmentDir);
                        repo.InsertAttachment(attachmentRecord);
                    }
                    catch
                    {
                        // Skip attachments that cannot be saved
                    }
                }

                result.Imported++;
                onProgress?.Invoke(result.Imported);
            }
            catch
            {
                result.Errors++;
            }
        });

        return result;
    }
}

// ── Abstractions for PST reading ─────────────────────────────────────────────

/// <summary>Implement this interface with your preferred PST/OST reading library.</summary>
public interface IPstReader
{
    Task ReadMessagesAsync(string pstFilePath, Action<IPstMessage> onMessage);
}

/// <summary>Represents a single email message from a PST/OST file.</summary>
public interface IPstMessage
{
    string? MessageId { get; }
    string? Subject { get; }
    string? SenderName { get; }
    string? SenderEmail { get; }
    string? DisplayTo { get; }
    string? DisplayCc { get; }
    string? DisplayBcc { get; }
    DateTime? DeliveryTime { get; }
    string? BodyText { get; }
    string? BodyHtml { get; }
    bool HasAttachments { get; }
    int AttachmentCount { get; }
    string? FolderPath { get; }
    IReadOnlyList<IPstAttachment> Attachments { get; }
}

/// <summary>Represents a single attachment from a PST/OST message.</summary>
public interface IPstAttachment
{
    string? Filename { get; }
    string? MimeType { get; }
    byte[]? Data { get; }
}

File: src/importer/pst-importer.ts
import path from 'path';
import { PSTFile, PSTFolder, PSTMessage, PSTAttachment } from 'pst-extractor';
import { PSTNodeInputStream } from 'pst-extractor/dist/PSTNodeInputStream.class';
import { Email } from '../models/email';
import { MailSearchRepository } from '../database/repository';
import { buildAttachment, parseAddressList, RawAttachment } from './email-normalizer';

export interface ImportResult {
  imported: number;
  skipped: number;
  errors: number;
}

/**
 * Import a PST/OST file into the MailSearch database.
 * Attachments are extracted to `attachmentDir`.
 */
export async function importPst(
  pstFilePath: string,
  repo: MailSearchRepository,
  attachmentDir: string,
  onProgress?: (imported: number) => void
): Promise<ImportResult> {
  const result: ImportResult = { imported: 0, skipped: 0, errors: 0 };

  const pstFile = new PSTFile(pstFilePath);
  const rootFolder = pstFile.getRootFolder();
  await processFolder(rootFolder, '', repo, pstFilePath, attachmentDir, result, onProgress);

  return result;
}

async function processFolder(
  folder: PSTFolder,
  folderPath: string,
  repo: MailSearchRepository,
  sourceFile: string,
  attachmentDir: string,
  result: ImportResult,
  onProgress?: (imported: number) => void
): Promise<void> {
  const currentPath = folderPath
    ? `${folderPath}/${folder.displayName}`
    : folder.displayName;

  if (folder.contentCount > 0) {
    let message = folder.getNextChild();
    while (message !== null) {
      if (message instanceof PSTMessage) {
        try {
          processMessage(message, currentPath, repo, sourceFile, attachmentDir, result);
          if (onProgress) {
            onProgress(result.imported);
          }
        } catch (_err) {
          result.errors++;
        }
      }
      message = folder.getNextChild();
    }
  }

  if (folder.hasSubfolders) {
    let subfolder = folder.getNextChild();
    while (subfolder !== null) {
      if (subfolder instanceof PSTFolder) {
        await processFolder(
          subfolder,
          currentPath,
          repo,
          sourceFile,
          attachmentDir,
          result,
          onProgress
        );
      }
      subfolder = folder.getNextChild();
    }
  }
}

function processMessage(
  message: PSTMessage,
  folderPath: string,
  repo: MailSearchRepository,
  sourceFile: string,
  attachmentDir: string,
  result: ImportResult
): void {
  const recipients = parseAddressList(message.displayTo);
  const ccRecipients = parseAddressList(message.displayCC);
  const bccRecipients = parseAddressList(message.displayBCC);

  const email: Email = {
    messageId: message.internetMessageId || undefined,
    subject: message.subject || '(no subject)',
    from: {
      name: message.senderName || undefined,
      email: message.senderEmailAddress || '',
    },
    to: recipients,
    cc: ccRecipients,
    bcc: bccRecipients,
    date: message.messageDeliveryTime || undefined,
    bodyText: message.body || undefined,
    bodyHtml: message.bodyHTML || undefined,
    hasAttachments: message.hasAttachments,
    attachmentCount: message.numberOfAttachments,
    sourceFile: path.resolve(sourceFile),
    importedAt: new Date(),
    folderPath,
  };

  const emailId = repo.insertEmail(email);
  if (emailId === 0) {
    // INSERT OR IGNORE returned 0 – duplicate message-id
    result.skipped++;
    return;
  }

  // Extract attachments
  if (message.hasAttachments) {
    for (let i = 0; i < message.numberOfAttachments; i++) {
      const attachment = message.getAttachment(i);
      if (!attachment || !attachment.filename) continue;

      const rawData = readPstAttachmentData(attachment);
      if (!rawData) continue;

      const raw: RawAttachment = {
        filename: attachment.longFilename || attachment.filename,
        mimeType: attachment.mimeTag || undefined,
        data: rawData,
      };

      try {
        const attachmentRecord = buildAttachment(emailId, raw, attachmentDir);
        repo.insertAttachment(attachmentRecord);
      } catch (_err) {
        // Skip attachments that cannot be saved
      }
    }
  }

  result.imported++;
}

function readPstAttachmentData(attachment: PSTAttachment): Buffer | null {
  const stream: PSTNodeInputStream | null = attachment.fileInputStream;
  if (!stream) return null;

  const size = attachment.filesize;
  const buffer = Buffer.alloc(size);
  stream.readCompletely(buffer);
  return buffer;
}
